sdk/src/client/worker.rs

```diff
@@ -0,0 +1,239 @@
+        let mut pool = self.0.lock().await;
+        let worker = pool.front().ok_or(WorkerError::EmptyPool)?.clone();
+        // Move the worker to the back
+        pool.rotate_left(1);
+        drop(pool);
+        let output = worker.process_task(priority, future).await?;
+        Ok(output)
+    }
+
+    pub(crate) async fn resize(&self, new_size: usize) -> Result<(), WorkerError> {
+        if new_size == 0 {
+            return Err(WorkerError::EmptyPool);
+        }
+        let mut pool = self.0.lock().await;
+        let curr_size = pool.len();
+        match new_size.cmp(&curr_size) {
+            core::cmp::Ordering::Less => {
+                while pool.len() > new_size {
+                    if let Some(worker) = pool.pop_front() {
+                        worker.exit()?;
+                    }
+                }
+            }
+            core::cmp::Ordering::Greater => {
+                while pool.len() < new_size {
+                    pool.push_front(Arc::new(Worker::spawn()));
+                }
+            }
+            core::cmp::Ordering::Equal => (),
+        }
+        Ok(())
+    }
+
+    pub(crate) async fn size(&self) -> usize {
+        self.0.lock().await.len()
+    }
+}
+
+pub(crate) enum WorkerEvent {
+    Task(WorkerTask),
+    Exit,
+}
+
+#[repr(u8)]
+#[derive(PartialEq, Eq, PartialOrd, Ord, Debug, Copy, Clone)]
+pub enum TaskPriority {
+    Low = 0,
+    Medium = 1,
+    High = 2,
+}
+
+#[derive(Debug)]
+pub(crate) struct Worker {
+    join_handle: JoinHandle<()>,
+    sender: tokio::sync::mpsc::UnboundedSender<WorkerEvent>,
+}
+
+impl Worker {
+    fn spawn() -> Self {
+        let (sender, mut recv) = tokio::sync::mpsc::unbounded_channel();
+        let join_handle = tokio::spawn(async move {
+            let mut queue = BinaryHeap::new();
+            let mut exiting = false;
+            // Wait to be awakened by the channel
+            while let Some(task) = recv.recv().await {
+                match task {
+                    WorkerEvent::Task(task) => queue.push(task),
+                    WorkerEvent::Exit => exiting = true,
+                }
+                loop {
+                    if !exiting {
+                        // Get up to 10 messages at a time
+                        for _ in 0..10 {
+                            match recv.try_recv() {
+                                Ok(task) => match task {
+                                    WorkerEvent::Task(task) => queue.push(task),
+                                    WorkerEvent::Exit => exiting = true,
+                                },
+                                Err(e) => match e {
+                                    TryRecvError::Empty => break,
+                                    TryRecvError::Disconnected => return,
+                                },
+                            }
+                        }
+                    }
+                    if let Some(next) = queue.pop() {
+                        next.await;
```

Comment:
Does it send any requests in parallel then?

Reply:
Yes, each worker is a parallel task, but the worker itself only processed in order.